query-service-impl/src/main/java/org/hypertrace/core/query/service/HandlerScopedMaskingConfig.java

```diff
@@ -0,0 +1,158 @@
+package org.hypertrace.core.query.service;
+
+import com.typesafe.config.Config;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.Value;
+import lombok.experimental.NonFinal;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class HandlerScopedMaskingConfig {
+  private static final String TENANT_SCOPED_MASKS_CONFIG_KEY = "tenantScopedMaskingCriteria";
+  private final Map<String, List<MaskValuesForTimeRange>> tenantToMaskValuesMap;
+  private HashMap<String, Boolean> shouldMaskAttribute = new HashMap<>();
+  private HashMap<String, String> maskedValue = new HashMap<>();
+
+  public HandlerScopedMaskingConfig(Config config) {
+    if (config.hasPath(TENANT_SCOPED_MASKS_CONFIG_KEY)) {
+      this.tenantToMaskValuesMap =
+          config.getConfigList(TENANT_SCOPED_MASKS_CONFIG_KEY).stream()
+              .map(maskConfig -> new TenantMasks(maskConfig))
+              .collect(
+                  Collectors.toMap(
+                      tenantFilters -> tenantFilters.tenantId,
+                      tenantFilters -> tenantFilters.maskValues));
+    } else {
+      this.tenantToMaskValuesMap = Collections.emptyMap();
+    }
+  }
+
+  public void parseColumns(ExecutionContext executionContext) {
+    shouldMaskAttribute.clear();
```

Comment:
It seems like the state is maintained per request, but we should only test against the timeRange condition.

Pre-compute using config:
- tenantId -> List of timeRanges
- timeRange -> set of attributes
- timeRange -> map(attributeId, maskValue)

During response processing:
- Check if any time range matches.
- Pick the first match (or should we apply UNION?).
   - If UNION is used, and the same attribute is present in two time ranges with different mask values, which one should we consider? I guess any Should be fine.

Reply:
This is how I've done it. 
In case of attribute in multiple time ranges, I choose any value.